src/rest_vol_dataset.c

```diff
@@ -5217,6 +5369,370 @@ RV_get_cmpd_subset_nmembers(hid_t src_type_id, hid_t dst_type_id, size_t *num_cm
     if (dst_member_type > 0)
         if (H5Tclose(dst_member_type) < 0)
             FUNC_DONE_ERROR(H5E_DATATYPE, H5E_CANTCLOSEOBJ, FAIL, "can't close destination datatype member");
+}
+
+/*-------------------------------------------------------------------------
+ * Function:    RV_vlen_data_to_json
+ *
+ * Purpose:     Converts a buffer of hvl_t instances to a JSON value
+ *              which may be sent to the server as part of a write.
+ *
+ *              Allocates memory under the *out pointer on success which must be
+ *              freed by the calling function.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ * Programmer:  Matthew Larson
+ *              January, 2024
+ */
+static herr_t
+RV_vlen_data_to_json(const void *_in, size_t nelems, hid_t dtype_id, void **out, size_t *out_size)
+{
+    herr_t       ret_value        = H5_ITER_CONT;
+    hid_t        parent_type      = H5I_INVALID_HID;
+    const hvl_t *in               = NULL;
+    H5T_class_t  parent_class     = H5T_NO_CLASS;
+    ptrdiff_t    elem_buf_ptrdiff = 0;
+
+    const char *const fmt_string          = "[%s]";
+    const char       *elem_fmt_string     = NULL;
+    char             *out_string          = NULL;
+    char             *out_string_curr_pos = NULL;
+    ptrdiff_t         out_diff            = 0;
+    int               bytes_printed       = 0;
+    size_t            tgt_out_size        = 0;
+    size_t            parent_dtype_size;
+
+    if (!out)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, H5_ITER_ERROR,
+                        "invalid NULL pointer provided for output of vlen to JSON conversion");
+
+    if (!_in)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, H5_ITER_ERROR,
+                        "invalid NULL pointer provided for input of vlen to JSON conversion");
+
+    in = (const hvl_t *)_in;
+
+    if ((parent_type = H5Tget_super(dtype_id)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_BADVALUE, FAIL, "can't get base type of vlen type");
+
+    if ((parent_class = H5Tget_class(parent_type)) < 0)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get class of parent of vlen dtype");
+
+    if ((parent_dtype_size = H5Tget_size(parent_type)) == 0)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get size of vlen seq base type");
+
+    if ((out_string = (char *)calloc(DATASET_VLEN_JSON_BODY_DEFAULT_SIZE, 1)) == NULL)
+        FUNC_GOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "can't allocate space for vlen data JSON");
+
+    *out_size           = DATASET_CREATION_PROPERTIES_BODY_DEFAULT_SIZE;
+    out_string_curr_pos = out_string;
+
+    /* Delineate start of list of sequences */
+    out_diff     = out_string_curr_pos - out_string;
```

Comment:
`out_string_curr_pos` is set to `out_string` directly above, so this and `tgt_out_size` below are always going to be 0 here. I don't think there's any need for these lines or the `CHECKED_REALLOC`.